### PR TITLE
atomic quadrature grid blocking [part 1] 

### PR DESCRIPTION
**Pull request by hokru**
## Description
This PR introduces an atomic blocking scheme for quadrature grid points. All grid points in a `BlockOPoints` object belong to a singular parent atom.
Part 1 contains just the basic feature to get it our for people depending on it. Optimization for speed is yet to come.

### New options:
- `DFT_BLOCK_SCHEME = ATOMIC` (unique blocks of grid points for each atom)
- `DFT_REMOVE_DISTANT_POINTS` (new flag for existing functionality)

### primary C++ feature:
- `grid->atomic_blocks()[N_ATOM][N_BLOCKS]` additionally to `grid->blocks()[N_BLOCKS]` # provides all grid points for an atom
- `block->parent_atom()` # atom the current block belongs to.

### fixes
- collocation size estimate in naive gridblocker was wrong.
-  removes unused `index` vector

#### note
- Automated formatting of the files with `clang-format` made unrelated changes.

## Questions
- [ ] The ugly code [here](https://github.com/hokru/psi4/blob/atomic_grid1/psi4/src/psi4/libfock/cubature.cc#L4630-L4641) is because of an issue with the `BlockOPoints` [object](https://github.com/hokru/psi4/blob/atomic_grid1/psi4/src/psi4/libfock/cubature.h#L179). I'd like advice how to handle this better.

## Checklist
- [x] Tests added for any new features

## Status
- [x] Ready for review
- [x] Ready for merge

---

## Usage

Normal loop structure
```
for (size_t Q = 0; Q < grid_->blocks().size(); Q++) {
  size_t parent_atom_ = grid_->blocks()[Q]->parent_atom();
  .
  .
  .
}
```

Looping over atoms and their blocks. 
```
# openmp note: for best performance the atom and block loop could possibly
# be collapsed into a singular loop using `collapse(2)`
for (size_t i = 0; i < grids_.size(); i++) { # here size = number of atoms
    for (size_t Q = 0; Q < grid_->atomic_blocks()[A].size(); Q++) {

        // access a block
        std::shared_ptr<BlockOPoints> block = grid_->atomic_blocks()[A][Q];
  .
  .
  .
}
```

### CODE REVIEW

**Review by JonathonMisiewicz — DISMISSED** [review]

File: psi4/src/psi4/libfock/cubature.cc [hunk 1]
```diff
@@ -4407,83 +4417,93 @@ MolecularGrid::~MolecularGrid() {
...
     z_ = blocker->z();
     w_ = blocker->w();
-    index_ = blocker->index();
 
     npoints_ = blocker->npoints();
     max_points_ = blocker->max_points();
     max_functions_ = blocker->max_functions();
     collocation_size_ = blocker->collocation_size();
-
+    size_t count = 0;
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libfock/cubature.cc:4456]: Was `count` an artifact of a previous version of this code? It doesn't seem to be doing anything.

File: psi4/src/psi4/libfock/cubature.cc [hunk 2]
```diff
@@ -4407,83 +4417,93 @@ MolecularGrid::~MolecularGrid() {
...
         blocks_.push_back(block[i]);
     }
+    if (options_.blockscheme == "ATOMIC") {
+        atomic_blocks_.resize(molecule_->natom());
+        for (size_t i = 0; i < molecule_->natom(); i++) {
+            for (size_t Q = 0; Q < blocker->atomic_blocks()[i].size(); Q++) {
+                atomic_blocks_[i].push_back(blocker->atomic_blocks()[i][Q]);
+            }
+        }
+    }
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libfock/cubature.cc:4469]: What's wrong with
```
blocks_ = blocker->blocks();
if (options_.blockscheme == "ATOMIC") {
    atomic_blocks_ = blocker->atomic_blocks();
}
```
?

I haven't tested this myself, but I've been [led to believe](https://herbsutter.com/2020/02/17/move-simply/) that should work. You might worry that the copying the objects is expensive, but "C++ already uses move automatically when copying from an object it knows will never be used again." Jet or Andy, let me know if I am misunderstanding move semantics.

File: psi4/src/psi4/libfock/cubature.cc [hunk 3]
```diff
@@ -4407,83 +4417,93 @@ MolecularGrid::~MolecularGrid() {
...
 }
 
 void MolecularGrid::remove_distant_points(double Rmax) {
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libfock/cubature.cc:4472]: Is this function tested anywhere?

File: psi4/src/psi4/libfock/cubature.h [hunk 4]
```diff
@@ -91,14 +92,16 @@ class MolecularGrid {
...
     std::shared_ptr<Matrix> orientation_;
     /// Radial grids, per atom
-    std::vector<std::shared_ptr<RadialGrid> > radial_grids_;
+    std::vector<std::shared_ptr<RadialGrid>> radial_grids_;
     /// Spherical grids, per atom and radial point
-    std::vector<std::vector<std::shared_ptr<SphericalGrid> > > spherical_grids_;
-    /// index_[fast_index] = slow_index
-    int* index_;
+    std::vector<std::vector<std::shared_ptr<SphericalGrid>>> spherical_grids_;
+    /// Grid points, per atom
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libfock/cubature.h:98]: Please add a comment somewhere to say that every `MolecularGrid` should have `atomic_grids_`, but whether it has `atomic_blocks_` is up to what blocking scheme(s) the user asked for.

**Review by JonathonMisiewicz — COMMENTED** [review]
This looks _much_ better. I have a few comments before I can approve this. :+1: for getting rid of `index_` and `index_ref_`.

File: psi4/src/psi4/libfock/cubature.cc [hunk 5]
```diff
@@ -4407,83 +4417,93 @@ MolecularGrid::~MolecularGrid() {
...
-            new NaiveGridBlocker(npoints_, x_, y_, z_, w_, index_, max_points, min_points, max_radius, extents_));
-    } else if (options_.get_str("DFT_BLOCK_SCHEME") == "OCTREE") {
+            new NaiveGridBlocker(npoints_, x_, y_, z_, w_, max_points, min_points, max_radius, extents_));
+    } else if (options_.blockscheme == "OCTREE") {
         blocker = std::shared_ptr<GridBlocker>(
-            new OctreeGridBlocker(npoints_, x_, y_, z_, w_, index_, max_points, min_points, max_radius, extents_));
+            new OctreeGridBlocker(npoints_, x_, y_, z_, w_, max_points, min_points, max_radius, extents_));
+    } else if (options_.blockscheme == "ATOMIC") {
+        blocker = std::shared_ptr<GridBlocker>(new AtomicGridBlocker(npoints_, x_, y_, z_, w_, max_points, min_points,
+                                                                     max_radius, extents_, molecule_, atomic_grids_));
```
[the remark this answers is not in the document]
Comment (hokru) [reply psi4/src/psi4/libfock/cubature.cc:4433]: No. To attach an object to you need to use `new` (I think) or switch to `make_shared`. 

I think the switch is preferred as it reads cleaner:
```
std::shared_ptr<GridBlocker> blocker
blocker = std::make_shared<AtomicGridBlocker>(args)
````

File: psi4/src/psi4/libfock/cubature.cc [hunk 6]
```diff
@@ -4322,6 +4321,13 @@ void DFTGrid::buildGridFromOptions(std::map<std::string, int> int_opts_map,
     opt.nradpts = full_int_options["DFT_RADIAL_POINTS"];
     opt.nangpts = full_int_options["DFT_SPHERICAL_POINTS"];
     opt.weights_cutoff = options_.get_double("DFT_WEIGHTS_TOLERANCE");
+    opt.blockscheme = full_str_options["DFT_BLOCK_SCHEME"];
+    opt.remove_distant_points = options_.get_bool("DFT_REMOVE_DISTANT_POINTS");
```
[the remark this answers is not in the document]
Comment (hokru) [reply psi4/src/psi4/libfock/cubature.cc:4325]: It is intentional. The previous limitation concerning `DFT_REMOVE_DISTANT_POINTS` for `ATOMIC` was removed.

File: psi4/src/psi4/libfock/cubature.cc (hunk 2 above)
Comment (hokru) [reply psi4/src/psi4/libfock/cubature.cc:4469]: stackexchange suggests it's ok (https://stackoverflow.com/a/14977748). I will change it.

File: psi4/src/psi4/libfock/cubature.cc (hunk 3 above)
Comment (hokru) [reply psi4/src/psi4/libfock/cubature.cc:4472]: It's regression tested indirectly in `dft-pruning` where the number of grid points are checked. I will add a more explicit regression test but we have no correctness test.

File: psi4/src/psi4/libfock/cubature.h (hunk 4 above)
Comment (hokru) [reply psi4/src/psi4/libfock/cubature.h:98]: I tried to clarify the situation with atomic_grids/blocks in their respective comments.

File: psi4/src/psi4/libfock/cubature.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libfock/cubature.cc:4433]: Duh. `blocker = std::make_shared<AtomicGridBlocker>(args)` is what I meant to type. If you can go with your switch, please do.

File: psi4/src/psi4/libfock/cubature.cc (hunk 1 above)
Comment (hokru) [reply psi4/src/psi4/libfock/cubature.cc:4456]: Not sure anymore. Removed now.

**Review by JonathonMisiewicz — APPROVED** [review]
As somebody new to the DFT code, LGTM. Thanks for the PR! It looks like this will enable some useful things.

**Review by jturney — APPROVED** [review]
This looks good!

**Review by maxscheurer — APPROVED** [review]
LGTM. 👍 for the tests!

**Pull request merged**